server/handlers.go

```diff
@@ -716,6 +721,27 @@ func (s *Server) handleAuthCode(w http.ResponseWriter, r *http.Request, client s
 		return
 	}
 
+	// Check for code challenge validity
+	if authCode.CodeChallenge != "" {
+		if verifier == "" {
+			s.tokenErrHelper(w, errAccessDenied, "Require code_verifier", http.StatusUnauthorized)
+			return
+		}
+		var challenge string
+		switch strings.ToLower(authCode.CodeChallengeMethod) {
+		case "s256":
+			sum := sha256.Sum256([]byte(verifier))
+			challenge = base64.RawURLEncoding.EncodeToString(sum[:])
+		// default to plain: insecure
+		default:
+			challenge = verifier
+		}
+		if challenge != authCode.CodeChallenge {
+			s.tokenErrHelper(w, errAccessDenied, "code_verifier doesn't have the same hash as code_challenge", http.StatusUnauthorized)
```

Comment:
According to [spec](https://tools.ietf.org/html/rfc7636#section-4.6), it's invalid_grant instead of access_denied.